fix(v1/kv_cache): resolve async KV transfer bug in cascade attention

ayushsatyam146 opened the pull request:
## Purpose
Solves #23130. This change fixes a critical bug in vLLM's cascade attention optimization in the V1 arch. The bug is in `get_num_common_prefix_blocks()`, which determines how many KV cache blocks are shared among all currently running requests to enable cascade attention optimizations.

## Changes made

* Replace ref_cnt-based common prefix detection with running request tracking
* Update get_num_common_prefix_blocks() to accept running_request_ids set
* Fix FullAttentionManager to count actual references from running requests
* Prevent incorrect cascade attention when async KV offloading delays cleanup

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -281,11 +280,17 @@ def remove_skipped_blocks(self, request_id: str,
-            if block.ref_cnt == num_running_requests:
+            actual_running_refs = sum(1 for req_id in running_request_ids
+                                      if req_id in self.req_to_blocks
+                                      and block in self.req_to_blocks[req_id])
```
`block in self.req_to_blocks[req_id]`  is an O(num_block) operation, making this function pretty slow.

ayushsatyam146 replied:
@heheda12345 thanks for the coments I am looking into it, but can you also please suggest something to make it optimal?

heheda12345 replied:
I'm thinking of
```
    def get_num_common_prefix_blocks(self, running_request_id: str,
                                     num_running_requests: int, transfering_request_ids: list[str]) -> int:
        running_blocks = self.req_to_blocks[running_request_id]
        transfering_blocks = [self.req_to_blocks[req_id] for req_id in transfering_request_ids]
        num_common_blocks = 0
        for i, block in enumerate(running_blocks):
            num_transfering_blocks = sum(1 for blocks in transfering_blocks if blocks[i].block_id == running_blocks[i].block_id)
            if block.ref_cnt == num_running_requests + num_transfering_blocks:
                num_common_blocks += 1
            else:
                break
```

heheda12345 reviewed (commented):
Even after `block in self.req_to_blocks[req_id]` is fixed, I'm still concern about the performance when all requests are sharing a very long prefix. The time complexity is num_requests x num_blocks_per_request. What about passing in the requests that are not running but are during kv transfer?

ayushsatyam146 commented:
Hi @heheda12345  @njhill  The time complexity of the new code is O(RxB) now, which was O(RxB²) in the previous iteration. I have one caching based implementation as well in mind which will bring down the complexity to O(1) best case and O(RxB) worst case. But that makes the code a little complex for this module hence I did not want to push that version without someone's approval. PTAL if this is fine or if we need to improve this further? Thanks!

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -281,15 +269,30 @@ def remove_skipped_blocks(self, request_id: str,
+            self, running_request_id: str, num_running_requests: int,
+            transfering_request_ids: list[str]) -> int:
+        """Get common prefix blocks using ref_cnt with transferring requests."""
+        if running_request_id not in self.req_to_blocks:
```
Will this happen?

ayushsatyam146 replied:
In majority of the cases it should not happen. I just put this as a safeguard since it won't cost much to put this check and although quite unlikely but there might be race conditions during request pre-emption and block allocation exceptions. If you feel like I should remove this please let me know and I will remove this. Thanks!

heheda12345 replied:
I prefer to not add code for cases that you feel won't happen. Otherwise we need to maintain the code in future updates.

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -281,15 +269,30 @@ def remove_skipped_blocks(self, request_id: str,
+        for i, block in enumerate(running_blocks):
+            num_transfering_blocks = sum(
+                1 for blocks in transfering_blocks if i < len(blocks)
+                and blocks[i].block_id == running_blocks[i].block_id)
```
```suggestion
                and blocks[i].block_id == block.block_id)
```

heheda12345 commented:
@ayushsatyam146 Hi, can you help to update this PR?

ayushsatyam146 commented:
Hi @heheda12345 sorry I got sick this week and couldn't work on this. But I am good now and will update this soon, Thanks for the patience.

ayushsatyam146 commented:
@heheda12345 gentle reminder to PTAL, thanks!

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -289,15 +277,34 @@ def remove_skipped_blocks(self, request_id: str,
-            if block.ref_cnt == num_running_requests:
+        for i, ref_block in enumerate(reference_blocks):
+            requests_with_block = sum(
+                1 for blocks in request_blocks if i < len(blocks)
```
please only iterate over transferring requests. Iterate over running requests is slow.

ayushsatyam146 replied:
If we don't iterate over running requests, then I think it's a little hard to ensure the complete logical correctness of this flow. I may be wrong here, but in my opinion, the below should work in most cases and I have done the same in the new changes that I pushed.
- If ref_block.ref_cnt >= total_requests (at least total_requests reference it)
- And all transferring requests have it
- Then, all running requests must also have it

Please correct me if there is a flaw in my logic or understanding of the flow. I will update the code changes accordingly. Thanks, and please take a look at this @heheda12345

ayushsatyam146 commented:
@heheda12345 Can you please take a look at this, Thanks!

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -292,15 +280,37 @@ def remove_skipped_blocks(self, request_id: str,
-            else:
+        for i, ref_block in enumerate(reference_blocks):
+
+            if ref_block.ref_cnt < total_requests:
```
is this correct? I think if a block is a common block for all running request, but not a common block for transferring request, we should regard it as a common block.

ayushsatyam146 replied:
yup this actually makes sense!

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -292,15 +280,37 @@ def remove_skipped_blocks(self, request_id: str,
+            if ref_block.ref_cnt < total_requests:
+                break
+
+            transferring_has_block = sum(
+                1 for blocks in transferring_blocks if i < len(blocks)
+                and blocks[i].block_id == ref_block.block_id)
+
+            if transferring_has_block != len(transfering_request_ids):
```
I'm thinking of:
```
if ref_block.ref_cnt - transferring_has_block == len(running_request_id): 
    # it is a common block
else:
    # it is not a common block
```
is it correct?

ayushsatyam146 replied:
Hi @heheda12345 I think this makes sense. It is also simpler and there is no significant performance overhead from this. I have implemented this in the new commit I pushed

ayushsatyam146 commented:
Hi @heheda12345 I resolved the merge conflicts on this and also included the changes suggested by you. Please take a look, Thanks!

heheda12345 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -597,7 +612,10 @@ def remove_skipped_blocks(self, request_id: str, num_computed_tokens: int) -> No
-        self, request_id: str, num_running_requests: int
+        self,
+        running_request_id: str,
+        running_request_ids: list[str],
```
```suggestion
        num_running_requests: int
```
Do we only need this?

ayushsatyam146 replied:
@heheda12345 I made this small update to the logic now, You can review it.

heheda12345 reviewed (commented):
This implementation looks great!

ayushsatyam146 commented on vllm/v1/core/single_type_kv_cache_manager.py:
```diff
@@ -303,15 +293,34 @@ def remove_skipped_blocks(self, request_id: str, num_computed_tokens: int) -> No
-        self, request_id: str, num_running_requests: int
+        self,
+        running_request_id: str,
+        running_request_ids: list[str],
+        transfering_request_ids: list[str],
     ) -> int:
-        blocks = self.req_to_blocks[request_id]
+        """Get common prefix blocks shared by all running requests."""
+
+        reference_blocks = self.req_to_blocks[running_request_id]
+
+        transferring_blocks = [
+            self.req_to_blocks[req_id]
+            for req_id in transfering_request_ids
+            if req_id in self.req_to_blocks
+        ]
+
         num_common_blocks = 0
-        for block in blocks:
-            if block.ref_cnt == num_running_requests:
+        for i, ref_block in enumerate(reference_blocks):
+            transferring_has_block = sum(
+                1
+                for blocks in transferring_blocks
+                if i < len(blocks) and blocks[i].block_id == ref_block.block_id
+            )
+
+            if ref_block.ref_cnt - transferring_has_block == len(running_request_ids):
                 num_common_blocks += 1
```
@heheda12345 I pointed this out in my previous implementations. 😅
Hence, I was explicitly iterating over both running and transferring requests but that will obviously be a performance bottleneck. IMO what we have right now is a good approximation. Please let me know if the current implementation is fine or not, Thanks!

heheda12345 replied:
I thought a block holding a reference should either be a running request or a transferring request. Are there any other cases?

And if there are some other cases, I think we can add these requests to transferring request (may need to change its name though)

heheda12345 replied:
Oh I see. There may be some requests that finish transferring and are changed to WAITING, but they are not scheduled in this step so it is not in RUNNING. Let me think about it. Will reply soon.

heheda12345 replied:
I think in this case, the current implementation will lead to incorrect behavior. E.g., some requests are holding the reference due to kv transfer, but in waiting state, and some other requests are in running state but don't have the common prefix.

What about just check `ref_cnt == len(self.req_to_blocks)`? (then we are finding the common prefix of all requests with KV cache, including the running, the transferring, and the ones that finishe transferring and are in waiting state).

heheda12345 commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -346,46 +346,30 @@ def reset_prefix_cache(self) -> bool:
-        NOTE(woosuk): The number of requests in the RUNNING state is **greater
-        than or equal to** the number of requests scheduled in the current step.
-        This is because the RUNNING state only indicates that:
-        1. The request has not yet finished, and
-        2. The request holds its blocks unfreed.
-
-        While all scheduled requests must be in the RUNNING state, the inverse
-        is not necessarily true. There may be RUNNING requests that are not
-        scheduled in the current step.
+        in the RUNNING state for each kv cache group. A block is considered a
+        common prefix block if it is referenced by ALL currently running
+        requests.
 
-        This can result in an edge case where the number of common prefix blocks
-        is 0, even though all scheduled requests share a common prefix. This
-        occurs because there may be unscheduled RUNNING requests that do not
-        share the common prefix. Currently, this case cannot be easily detected,
-        so the function returns 0 in such cases.
```
I think we still need to keep this comment about the mismatch between scheduled request and running request. For example, in pipeline parallel, there are two stages, requests in both stages are running but only requests in the current stage is scheduled.

ayushsatyam146 replied:
Okay sure

heheda12345 commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -346,46 +346,30 @@ def reset_prefix_cache(self) -> bool:
-                state. This can be different from the number of scheduled
-                requests in the current step.
+            running_request_id: The request ID of the running request.
+            num_running_requests: The number of requests in the RUNNING state.
```
similar, can you revert this comment?

ayushsatyam146 committed:
fix(v1/kv_cache): resolve async KV transfer bug in cascade attention
* Replace ref_cnt-based common prefix detection with running request tracking
* Update get_num_common_prefix_blocks() to accept running_request_ids set
* Fix FullAttentionManager to count actual references from running requests
* Prevent incorrect cascade attention when async KV offloading delays cleanup

This resolves a bug where completed requests with pending async transfers
still contributed to ref_cnt, causing incorrect cascade attention decisions.

Signed-off-by: Ayush Satyam <ayushsatyam146@gmail.com>

ayushsatyam146 commented:
Hi @heheda12345, I went through this approach, and apart from some occasional conservative handling of cascade attention, it looks good overall. I’ve implemented it as well — please take a look when you get a chance. Thanks!

heheda12345 reviewed (approved):
LGTM! I think this solution is clean.
